designs/Compiler B: Built-in libraries-design.md

```diff
@@ -0,0 +1,161 @@
+# 10 Test Cases
+
+Sadly, all we can do right now are simply math functions. Other builtin functions, like reduce, map and filter, are related to the implmentation of iterable Object, Generics and polymorphism. So maybe we will write these high level functions later.
+
+### Factorial
+
+input: 
+c : int = 4
+print(factorial(c))
+Output: 
+24
+
+### Max
+
+input: 
+c : int = 3
+d: int = 2
+print(max(factorial(c), factorial(d)))
+Output: 
+6
+
+### Pow
+
+input: 
+c : int = 3
+d: int = 2
+print(pow(c, d))
+Output: 
+9
+
+### Randint
+
+input: 
+a : int = 0
+b : int = 5
+print(randint(a,b))
+Output: 
+A random integer between 0 and 5
+
+### Randrange
+
+input: 
+a : int = 0
+b: int = 101
+print(randrange(0, 101, 2))
+Output: 
+A random even integer from 0 to 100
+
+### Time
+
+input: 
+print(time())
+Output: 
+1436428326(the current time stamp)
+
+### LCM (least common multiple)
+
+input: 
+print(lcm(4,6))
+print(lcm(3,5))
+print(lcm(2,8))
+Output: 
+12
+15
+8
+
+### GCD (greatest common divisor)
+
+input: 
+print(gcd(4,6))
+print(gcd(3,5))
+print(gcd(2,8))
+Output: 
+2
+1
+2
+
+### Combination
+
+input: 
+print(comb(4,2))
+print(comb(10,3))
+print(comb(20,5))
+Output: 
+6
+120
+15504
+
+### Perm
+
+input: 
+print(perm(4,2))
+print(perm(10,3))
+print(perm(20,5))
+Output: 
+12
+720
+1860480
+
+# AST/IR changes to make
+
+Since we will have to high-level functions, we suggest to add another basic type function like below in ast.ts
+
+Type = 
+
+​	|... 
+
+​	| {
+
+​		tag: "function",
```

Comment:
Interesting idea, why define function as a new type? are you planning to parse function as a type in general ( e.g. class<function>, etc.)?

Reply:
I give up this idea.